### DISCUSSION TITLE
A class instance as a coroutine

### DISCUSSION THREAD

**Discussion by davidbrochart** [agronholm/anyio#759] [Q&A] [answered]
I'd like to have a class instance look like a coroutine:
```py
class Foo:
    def __await__(self):
        return self.run().__await__()

    async def run(self):
        print("run")
```
But [AnyIO doesn't recognize it as a coroutine](https://github.com/agronholm/anyio/blob/e7f750b96f5416d8ae932e15d726b5d03de80b67/src/anyio/_backends/_asyncio.py#L752):
```py
import anyio

async def main():
    async with anyio.create_task_group() as tg:
        tg.start_soon(Foo)

anyio.run(main)

# TypeError: Expected __main__.Foo() to return a coroutine, but the return value (<__main__.Foo object at 0x72557fdc28a0>) is not a coroutine object
```
Is there a better way to write my class so that AnyIO accepts it as a task function?

**Comment by uSpike** [comment]
Not sure if your example has a typo or not, but you're passing the Foo class and not an instance.  You have to pass an instance of Foo.

**Reply by davidbrochart** [reply]
Yes it's intended to pass `Foo`, since `Foo()` is a coroutine (it can be awaited). `start_soon` accepts a coroutine function and will call it to make it a coroutine, right?

**Reply by smurfix** [reply]
`Foo` is primarily a class and `bar=Foo()` an object of that class. The fact that this object is also an Awaitable (you can say `await bar`) doesn't make `Foo` a coroutine.

That being said, fake it till you make it:

``` python3
import typing

class Foo(typing.Coroutine):
    def __await__(self):
        return self.run().__await__()

    def send(self, value=None):
        return self.__await__().send(value)

    def throw(self, *_unused):
        raise RuntimeError("Shouldn't happen")

    async def run(self):
        print("run")
```

**Reply by smurfix** [reply]
though frankly I'd recommend being somewhat more explicit and using `tg.start_soon(Foo().run)` instead.

**Reply by davidbrochart** [reply]
Thanks @smurfix, strangely your `Foo` implementation still doesn't work on AnyIO/asyncio, but it works on Trio:
```py
import anyio

async def main():
    async with anyio.create_task_group() as tg:
        tg.start_soon(Foo)

anyio.run(main, backend="trio")
```

**Reply by smurfix** [reply]
Huh. Works for me, yes also with asyncio backend. Python 3.11.2. What's your version (and your error backtrace)?

**Reply by davidbrochart** [reply]
I tried with Python 3.11.2, same error:
```
Traceback (most recent call last):
  File "/home/david/git/anycorn/foo.py", line 24, in main
    tg.start_soon(Foo)
  File "/home/david/micromamba/envs/anycorn/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 777, in start_soon
    self._spawn(func, args, name)
  File "/home/david/micromamba/envs/anycorn/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 754, in _spawn
    raise TypeError(
TypeError: Expected __main__.Foo() to return a coroutine, but the return value (<__main__.Foo object at 0x73cc37b92f50>) is not a coroutine object

During handling of the above exception, another exception occurred:

  + Exception Group Traceback (most recent call last):
  |   File "/home/david/git/anycorn/foo.py", line 27, in <module>
  |     anyio.run(main, backend="asyncio")
  |   File "/home/david/micromamba/envs/anycorn/lib/python3.11/site-packages/anyio/_core/_eventloop.py", line 74, in run
  |     return async_backend.run(func, args, {}, backend_options)
  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |   File "/home/david/micromamba/envs/anycorn/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 2034, in run
  |     return runner.run(wrapper())
  |            ^^^^^^^^^^^^^^^^^^^^^
  |   File "/home/david/micromamba/envs/anycorn/lib/python3.11/asyncio/runners.py", line 118, in run
  |     return self._loop.run_until_complete(task)
  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |   File "/home/david/micromamba/envs/anycorn/lib/python3.11/asyncio/base_events.py", line 653, in run_until_complete
  |     return future.result()
  |            ^^^^^^^^^^^^^^^
  |   File "/home/david/micromamba/envs/anycorn/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 2022, in wrapper
  |     return await func(*args)
  |            ^^^^^^^^^^^^^^^^^
  |   File "/home/david/git/anycorn/foo.py", line 22, in main
  |     async with anyio.create_task_group() as tg:
  |   File "/home/david/micromamba/envs/anycorn/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 680, in __aexit__
  |     raise BaseExceptionGroup(
  | ExceptionGroup: unhandled errors in a TaskGroup (1 sub-exception)
  +-+---------------- 1 ----------------
    | Traceback (most recent call last):
    |   File "/home/david/git/anycorn/foo.py", line 24, in main
    |     tg.start_soon(Foo)
    |   File "/home/david/micromamba/envs/anycorn/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 777, in start_soon
    |     self._spawn(func, args, name)
    |   File "/home/david/micromamba/envs/anycorn/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 754, in _spawn
    |     raise TypeError(
    | TypeError: Expected __main__.Foo() to return a coroutine, but the return value (<__main__.Foo object at 0x73cc37b92f50>) is not a coroutine object
```
It boils down to `inspect.iscoroutine(Foo())` being `False`, instead of `True`.

**Reply by smurfix** [reply, marked as answer]
Sorry about this; I had patched out that test and my editor refused to restore the original without telling me about it. *Sigh*.

asyncio is not using `typing.Coroutine` for this test for the simple reason that its code is somewhat older than `typing`.

AFAIK unless the asyncio people accept a patch that changes this you're out of luck here.

**Reply by davidbrochart** [reply]
> though frankly I'd recommend being somewhat more explicit and using `tg.start_soon(Foo().run)` instead.

Thanks, I guess I'll be more explicit then :smile:

**Reply by davidbrochart** [reply]
> asyncio is not using `typing.Coroutine` for this test for the simple reason that its code is somewhat older than `typing`.

Actually, as I said it fails because AnyIO is using `inspect.iscoroutine(Foo())`, so it's not related to asyncio?
And the code for [inspect.iscoroutine](https://github.com/python/cpython/blob/bdab67e1c795443a0d8f8a5bbeb3a91ac4fd5a19/Lib/inspect.py#L463C1-L465C51) is:
```py
def iscoroutine(object):
    """Return true if the object is a coroutine."""
    return isinstance(object, types.CoroutineType)
```
But a class cannot inherit from `types.CoroutineType`, which is defined [here](https://github.com/python/cpython/blob/bdab67e1c795443a0d8f8a5bbeb3a91ac4fd5a19/Lib/types.py#L30-L32) as:
```py
async def _c(): pass
_c = _c()
CoroutineType = type(_c)
_c.close()  # Prevent ResourceWarning
```

**Reply by smurfix** [reply]
Yeah, you're right, I seem to have gotten confused. s/asyncio/inspect.

**Reply by davidbrochart** [reply]
Actually asyncio doesn't use `inspect.iscoroutine(Foo())`, only AnyIO does.
So this works with asyncio:
```py
import asyncio
import typing

class Foo(typing.Coroutine):
    def __await__(self):
        return self.run().__await__()

    def send(self, value=None):
        return self.__await__().send(value)

    def throw(self, *_unused):
        raise RuntimeError("Shouldn't happen")

    async def run(self):
        print("run")

async def main():
    foo = Foo()
    task = asyncio.create_task(foo)
    await task

asyncio.run(main())
```
And this works with Trio:
```py
import typing

import trio

class Foo(typing.Coroutine):
    def __await__(self):
        return self.run().__await__()

    def send(self, value=None):
        return self.__await__().send(value)

    def throw(self, *_unused):
        raise RuntimeError("Shouldn't happen")

    async def run(self):
        print("run")

async def main():
    async with trio.open_nursery() as nursery:
        nursery.start_soon(Foo)

trio.run(main)
```
But this doesn't work with AnyIO:
```py
import typing

import anyio

class Foo(typing.Coroutine):
    def __await__(self):
        return self.run().__await__()

    def send(self, value=None):
        return self.__await__().send(value)

    def throw(self, *_unused):
        raise RuntimeError("Shouldn't happen")

    async def run(self):
        print("run")

async def main():
    async with anyio.create_task_group() as tg:
        tg.start_soon(Foo)

anyio.run(main)
```

**Reply by davidbrochart** [reply]
I opened https://github.com/agronholm/anyio/pull/762.

**Reply by graingert** [reply]
> `Foo` is primarily a class and `bar=Foo()` an object of that class. The fact that this object is also an Awaitable (you can say `await bar`) doesn't make `Foo` a coroutine.
> 
> That being said, fake it till you make it:
> 
> ```python
> import typing
> 
> class Foo(typing.Coroutine):
>     def __await__(self):
>         return self.run().__await__()
> 
>     def send(self, value=None):
>         return self.__await__().send(value)
> 
>     def throw(self, *_unused):
>         raise RuntimeError("Shouldn't happen")
> 
>     async def run(self):
>         print("run")
> ```

@smurfix `        return self.__await__().send(value)` will just restart the `run(` coroutine every time send( is called and will never allow it to make progress.
